pkg/providers/tf/job_runner.go

```diff
@@ -0,0 +1,233 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tf
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/GoogleCloudPlatform/gcp-service-broker/brokerapi/brokers/models"
+	"github.com/GoogleCloudPlatform/gcp-service-broker/db_service"
+	"github.com/GoogleCloudPlatform/gcp-service-broker/pkg/providers/tf/wrapper"
+	"github.com/GoogleCloudPlatform/gcp-service-broker/utils"
+)
+
+const (
+	InProgress = "in progress"
+	Succeeded  = "succeeded"
+	Failed     = "failed"
+)
+
+// NewTfJobRunerFromEnv creates a new TfJobRunner with default configuration values.
+func NewTfJobRunerFromEnv() (*TfJobRunner, error) {
+	projectId, err := utils.GetDefaultProjectId()
+	if err != nil {
+		return nil, err
+	}
+
+	return &TfJobRunner{
+		ProjectId:      projectId,
+		ServiceAccount: utils.GetServiceAccountJson(),
+	}, nil
+}
+
+// TfJobRunner is responsible for executing terraform jobs in the background and
+// providing a way to log and access the state of those background tasks.
+//
+// Jobs are given an ID and a workspace to operate in, and then the TfJobRunner
+// is told which Terraform commands to execute on the given job.
+// The TfJobRunner executes those commands in the background and keeps track of
+// their state in a database table which gets updated once the task is completed.
+//
+// The TfJobRunner keeps track of the workspace and the Terraform state file so
+// subsequent commands will operate on the same structure.
+type TfJobRunner struct {
+	ProjectId      string
+	ServiceAccount string
+}
+
+// StageJob stages a job to be executed. Before the workspace is saved to the
+// database, the modules and inputs are validated by Terraform.
+func (runner *TfJobRunner) StageJob(ctx context.Context, jobId string, workspace *wrapper.TerraformWorkspace) error {
+	// Validate that TF is happy with the workspace
+	if err := workspace.Validate(); err != nil {
+		return err
+	}
+
+	workspaceString, err := workspace.Serialize()
+	if err != nil {
+		return err
+	}
+
+	deployment := &models.TerraformDeployment{
+		ID:                jobId,
+		Workspace:         workspaceString,
+		LastOperationType: "validation",
+	}
+
+	return runner.operationFinished(nil, workspace, deployment)
+}
+
+func (runner *TfJobRunner) markJobStarted(ctx context.Context, deployment *models.TerraformDeployment, operationType string) error {
+	// update the deployment info
+	deployment.LastOperationType = operationType
+	deployment.LastOperationState = InProgress
+	deployment.LastOperationMessage = ""
+
+	if err := db_service.SaveTerraformDeployment(ctx, deployment); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (runner *TfJobRunner) hydrateWorkspace(ctx context.Context, deployment *models.TerraformDeployment) (*wrapper.TerraformWorkspace, error) {
```

Comment:
I used hydrate because we aren't really creating the workspace on disk, nor are we really creating a new one, but we are deserializing it and augmenting it with additional data.